Pull Request for Code Review

gmmowry opened the pull request:
Features in pull request: 
User can specify interest category - hard coded
User can create route between two points - hard coded
Seeded with yelp data - hard coded into database
Categories with subcategories are coded into database
Searches for attractions within a specified radius of the route polyline
Users route only shows attractions in the categories they have marked as their interests
User can see basic attraction info when they click on it
User can login/sign-up/logout

gmmowry committed:
models for category, attraction, user, trip and location created with fields and associations

lduan committed:
added timestamps to models

gmmowry committed:
Merge pull request #1 from lduan/master

added timestamps to models

lduan committed:
added some fixes for models

nelltaylor committed:
Merge pull request #2 from lduan/master

added some fixes for models

lduan committed:
further model corrections

lduan committed:
initial seed file created and is working

lduan committed:
added fake secret keys to travis.yml

nicole-a-tesla committed:
base mapping and ajax functionality

nicole-a-tesla committed:
Ajax and rails are now getting along. Json attraction objects now pass to view.

lduan committed:
only load the yelp.yml file in development

nelltaylor committed:
Merge pull request #3 from lduan/master

Fixed some models (typos) and added a working intial seed file.

nicole-a-tesla committed:
"Add Attraction" button now sends ajax request to pull 5 random attractions from teh db and generate markers for them on the map. Woo!

nicole-a-tesla committed:
Merge pull request #1 from nicole-a-tesla/map-making

Map making

nelltaylor committed:
Generated spec directory and files, added gems to gemfile including rspec, underscore-rails, polylines for decoding/encoding, and simplecov.

nelltaylor committed:
Merge pull request #4 from nicole-a-tesla/master

ajax marker generation for random attractions

nelltaylor committed:
Merged Nicole's changes with Gemfile updates including rails-rspec, rails-underscore, 12 factor, simple cov, polylines

nelltaylor committed:
Reconciled gemfile.lock

nelltaylor committed:
Fixed Gemfile.lock merge conflicts again

nelltaylor committed:
Merge pull request #7 from nelltaylor/master

Reconciled Gemfile.lock  merge

nicole-a-tesla committed:
siloed google api key into gitignored yml file, added a yml.example file

nicole-a-tesla committed:
Merge pull request #2 from nicole-a-tesla/map-making

siloed google api key into gitignored yml file, added a yml.example file

nicole-a-tesla committed:
took comments and logs out of attractions.js

nicole-a-tesla committed:
Changed base route to chicago to st. louis. Stripped out unused variables in mapOptions.

nicole-a-tesla committed:
Merge pull request #3 from nicole-a-tesla/map-making

changed base route to Chi --> St. Louis

gmmowry committed:
Merge pull request #9 from nicole-a-tesla/master

changed base route to Chi --> St. Louis

nelltaylor committed:
Reformmated Attraction seed and model to support GeoJSON formatting for coordinates. Working.

gmmowry committed:
Merge pull request #10 from nelltaylor/master

Reformmated Attraction seed and model to support GeoJSON ...

gmmowry committed:
added decoding for polyline and inversion for lat and long coordinates

gmmowry committed:
added polyline.js library

nelltaylor committed:
Merge pull request #11 from gmmowry/master

Added polyline.js library, decode for polyline added, and swap [lat,lon] to [lon,lat]

nicole-a-tesla committed:
added turf.js

nicole-a-tesla committed:
polygon gerJson object gets sent to /maps/search

nicole-a-tesla committed:
Merge pull request #4 from nicole-a-tesla/polygon

Polygon

gmmowry committed:
geoWithin is not working for some reason

nicole-a-tesla committed:
removed unnecessary comment

nicole-a-tesla committed:
Merge pull request #5 from nicole-a-tesla/polygon

removed unnecessary comment

lduan committed:
Merge pull request #12 from nicole-a-tesla/master

polygon generates and is sent to /maps/search

lduan committed:
Merge branch 'master' of github.com:bison-2014/laika into geoWithin

lduan committed:
attraction model now can search for attractions within a polygon geoJSON

nelltaylor committed:
Merge pull request #13 from bison-2014/geoWithin

Added geowithin search in Attractions model

nelltaylor committed:
Fixed data type issues in ajax request in map.js and maps_controller.rb, properly returning objects for search_within query.

nelltaylor committed:
Rendered attractions on the map, fixed loadMarkers method, changed coordinates to new longlat format in attractions.js and maps.js.

nicole-a-tesla committed:
polygon is now drawn on map on route calculation

gmmowry committed:
added mongo-rspec gem, validations for attractions, location and category

gmmowry committed:
validation tests for user and location models created

nicole-a-tesla committed:
Merge pull request #14 from nelltaylor/master

Fixed data type issues in ajax request in map.js and maps_controller.rb

nicole-a-tesla committed:
Merge pull request #6 from nicole-a-tesla/draw-polygon

Draw polygon

gmmowry committed:
Merge pull request #15 from nicole-a-tesla/master

Draw polygon

gmmowry committed:
added one user seed for testing purposes

nicole-a-tesla committed:
Merge pull request #16 from gmmowry/master

Added validations for all models, validate spec tests written and passing, added bcrypt and mongoid-rspec gem

nelltaylor committed:
Began adding Foundation framework

nelltaylor committed:
Merged updates to models and seeds with Foundation framework generation

nicole-a-tesla committed:
when markers are clicked on, info windows pop up displaying the attraction's name

nicole-a-tesla committed:
Merge pull request #7 from nicole-a-tesla/info-windows

Info windows

gmmowry committed:
Merge pull request #17 from nicole-a-tesla/master

Info windows pop up on marker click

nelltaylor committed:
Foundation working, need to style

nelltaylor committed:
Merging maps.js updates with Foundation styling.
Merge branch 'master' of https://github.com/bison-2014/laika

gmmowry committed:
corrected model associations and added user seed

nelltaylor committed:
Merge pull request #18 from gmmowry/master

corrected model associations and added user seed

nelltaylor committed:
Merging model updates.
Merge branch 'master' of https://github.com/bison-2014/laika

gmmowry committed:
Merge pull request #19 from nelltaylor/master

Added Foundation Framework

nelltaylor committed:
Resize map dynamically in map.js.

nicole-a-tesla committed:
Merge pull request #20 from nelltaylor/master

Resize map dynamically in map.js.

nicole-a-tesla committed:
fixed indentation

lduan committed:
fixed a issue with attraction models (yelp_id is a string, not an integer).  Corrected seeds file to Attraction.find_or_initialize_by instead of Attraction.new.

gmmowry committed:
Merge pull request #21 from lduan/master

fixed a issue with attraction models (yelp_id is a string, not an integer).  Corrected seeds file to Attraction.find_or_initialize_by instead of Attraction.new.

nicole-a-tesla committed:
Route is now an object and renders to page when told to calculate itself. No polygon functionality yet.

gmmowry committed:
minor changes to user model for validation editing, potential category search filter added but need current_user data

nelltaylor committed:
Set up and tested authentication, sign up, and log out for users and sessions, added BSON serializer to initialization.

gmmowry committed:
Merge pull request #22 from nelltaylor/master

Set up and tested authentication, sign up, and log out

gmmowry committed:
added simplecov in rails_helper for testing

gmmowry committed:
added in category search filter

lduan committed:
fixed coordinate retrieval (#latitude and #longitude) in attraction and location

lduan committed:
updated seeds to seed location, categories, and then all attractions for each.  A new model 'Seed' is needed to keep a record of what has been seeded and what hasn't been

lduan committed:
fixed location spec

nelltaylor committed:
View begun for index, set for desktop display. Working on form to submit origin and destination to Google Maps.

gmmowry committed:
Merge pull request #23 from lduan/master

Added code to seeds that will request businesses for each location and category we want.

nelltaylor committed:
Seeds working.
Merge branch 'master' of https://github.com/bison-2014/laika

gmmowry committed:
edited readme with team names and some instructions

gmmowry committed:
in current_user help switched find to find_by

gmmowry committed:
added in category filter for search_within method

nelltaylor committed:
Merge pull request #24 from gmmowry/master

Added category filter based on user interests for search, added simplecov lines to rails_helper, edited readme file

nicole-a-tesla committed:
object orientified map.js

nicole-a-tesla committed:
object orientified map.js

nicole-a-tesla committed:
Merge pull request #8 from nicole-a-tesla/object-orientify-js

Object orientify js

nelltaylor committed:
Merge pull request #25 from nicole-a-tesla/master

oojs

nelltaylor committed:
Merged pull request for categories filtering

nelltaylor committed:
Merging map.js updates.
Merge branch 'master' of https://github.com/bison-2014/laika

nelltaylor committed:
Added first views, added form to search for origin and destination. Not linked yet.

gmmowry committed:
Merge pull request #26 from nelltaylor/master

Views begun, form to connect origin and destination set up but not linked to map.js yet.

lduan committed:
updated config api keys reading to be compatible with heroku

gmmowry committed:
created contributing file, added pull request and code review workflows

gmmowry committed:
added heroku link, links to team member github pages, and instructions to readme

gmmowry committed:
removed simplecov from rails_helper

gmmowry committed:
added ratings and reviews to infowindows for attractions

nicole-a-tesla committed:
Merge pull request #27 from gmmowry/master

Edited ReadMe, created Contributing file, removed simplecov issue, added ratings and reviews to infowindows

gmmowry committed:
commented out code to have categories for each attraction showing on the infoWindow, not working yet

gmmowry committed:
added more attributes to Attraction seeds and fixed seeding order of models

gmmowry committed:
added in first yelp_category to infowindow for attractions

gmmowry committed:
added in main categories for active life, hotels and travel, and nightlife and all subcategories

gmmowry committed:
removed commented out unecessary method code in attraction.rb

nicole-a-tesla committed:
Merge pull request #29 from gmmowry/master

Updated seeds file, changed origin/destination to cities in db, added categories to attractions' infowindows and removed unnecessary code

gmmowry committed:
added comments in for our chicago/seattle demo roadtrip, no code touched

gmmowry committed:
links to businesses yelp pages working and opening in a new tab

nicole-a-tesla committed:
list of attractions dynamically added to bottom of screen

nicole-a-tesla committed:
Merge pull request #9 from nicole-a-tesla/ajaxy-attraction-lists

Ajaxy attraction lists

gmmowry committed:
Merge pull request #30 from nicole-a-tesla/master

jquery attraction lists

gmmowry committed:
notes/comments added for demo seeding

gmmowry committed:
edited instructions

lduan committed:
fixed bugs

lduan committed:
heroku fix

rmw commented on app/assets/javascripts/attraction_list.js:
```diff
@@ -0,0 +1,7 @@
+// $(document).ready(function(){
+  var loadAttractionList = function(attraction, marker){
+
+    var attractionInfo = ['<li id="attraction', attraction._id + '">',attraction.name,'</li>'].join('')
```
Take a look at [Underscore.js templates](http://underscorejs.org/#template) for a bit of a nicer way to handle this.

rmw commented on app/assets/javascripts/attractions.js:
```diff
@@ -0,0 +1,39 @@
+$(document).ready(function(){
```
:+1: to separate files!!

rmw commented on app/assets/javascripts/attractions.js:
```diff
@@ -0,0 +1,39 @@
+    $.ajax({
+      url: url,
+      type: "get",
+      beforeSend: function(xhr) { xhr.setRequestHeader('X-CSRF-Token', $('meta[name=csrf-token]').attr('content'))},
```
Assuming you have multiple ajax requests, I'd recommend you use this (once) instead:

``` javascript
$.ajaxSetup({
  headers: {
    'X-CSRF-Token': $('meta[name=csrf-token]').attr('content')
  }
});
```

rmw commented on app/assets/javascripts/attractions.js:
```diff
@@ -0,0 +1,39 @@
+  var marker = new google.maps.Marker({
+    position: coords,
+    map: map
+  });
```
I'd love to see this written OO style.

Perhaps you'd have a `Marker` constructor function object and a `MarkerCollection` object.  Clicking 'Add Attraction`would call`mc = new MarkerCollection(url)` which in turn would make an ajax call and then create a new Marker object.

rmw commented on app/assets/javascripts/attractions.js:
```diff
@@ -0,0 +1,39 @@
+      dataType: 'json'
+    }).done(function(data){
+        loadMarkers(data)
+    });
```
You should always handle `error` in JS as well, because from the user's perspective nothing is happening.

rmw commented on app/assets/javascripts/map.js:
```diff
@@ -0,0 +1,210 @@
+
+var directionsService = new google.maps.DirectionsService();
+
+function initialize() {
```
This should really be part of an object -- at the very least a method off an object literal -- so you don't have a very generically named method `initialize` in the global namespace just waiting to be overwritten. :smiling_imp:

rmw commented on app/assets/javascripts/map.js:
```diff
@@ -0,0 +1,210 @@
+  this.waypts = this.getWaypoints();
+};
+
+Route.prototype.getWaypoints = function(){
```
:+1: OO here!

rmw commented on app/assets/javascripts/map.js:
```diff
@@ -0,0 +1,210 @@
+
+Route.prototype.getWaypoints = function(){
+  var waypts = [];
+  var checkboxArray = document.getElementById('waypoints');
```
You might consider having a `RouteView` object that is responsible for accessing the DOM and translating that into data and a `Route` object that's responsible for calculating the route and holding data.

rmw commented on app/assets/javascripts/map.js:
```diff
@@ -0,0 +1,210 @@
+    travelMode: google.maps.TravelMode.DRIVING
+  };
+
+  directionsService.route(request, function(response, status){
```
To maintain encapsulation, it'd be better to pass the `directionsService` into the method or the constructor function than to use a global object here.

rmw commented on app/assets/javascripts/map.js:
```diff
@@ -0,0 +1,210 @@
+
+      decoder = new PolylineDecoder()
+      route.polyline = decoder.decodePolyline(response);
+      polygon = new Polygon(route.polyline)
```
Seems like to create this global `polygon` object, which isn't used anywhere (at least not in this scope).  Don't create global vars (use `var`).  If a method returns an object that you don't need, don't bother storing it in a variable.

rmw commented on app/assets/javascripts/map.js:
```diff
@@ -0,0 +1,210 @@
+}
+
+
+//----polyline decoder-----
```
Route and Polyline should really be in their own files.  With good encapsulation of objects this should totally be possible -- you'll have an `App` or other such object that acts as the controller and sets up the relationship between the objects during `$(document).ready`

rmw commented on app/assets/javascripts/map.js:
```diff
@@ -0,0 +1,210 @@
+
+//----polyline decoder-----
+
+var PolylineDecoder = function(response){}
```
Does this function need to take an argument?

rmw commented on app/assets/javascripts/map.js:
```diff
@@ -0,0 +1,210 @@
+
+Polygon.prototype.searchWithin = function(polygon){
+  $.ajax({
+  url: '/maps/search',
```
indentation!

rmw commented on app/assets/javascripts/map.js:
```diff
@@ -0,0 +1,210 @@
+  var coordsToDraw = this.createLatLongObjects(geoJsonObject)
+
+    bufferedPolygon = new google.maps.Polygon({
+    paths: coordsToDraw,
```
indentation!

rmw commented on app/assets/javascripts/map.js:
```diff
@@ -0,0 +1,210 @@
+
+//------------Markers----------------
+
+function loadMarkers(attractions){
```
These seems to be a duplicate of code in another file?

rmw commented on app/assets/javascripts/map.js:
```diff
@@ -0,0 +1,210 @@
+
+//-----------InfoBox----------------
+var InfoBox = function(attraction, marker){
+  this.contentString ='<div>' +
```
Generally, it's better practice to only store needed variables (attraction, marker) in the constructor function and then add an `init` method which does set up.  This prevents you from copying a lot of code between each object.  Or if this is a one-time only object, you might consider using an object literal.  

What do you think @m3rabb?

rmw replied:
Also, see note about underscore templates above.

rmw commented on app/assets/stylesheets/application.css.scss:
```diff
@@ -12,4 +12,45 @@
+
  */
+
+html body {
```
Generally we try not to put CSS in this particular file and instead put this sort of stuff in a `_base.scss` or `_layout.scss` file that gets required first (or close to first).

rmw commented on app/assets/stylesheets/application.css.scss:
```diff
@@ -12,4 +12,45 @@
  *
  *= require_tree .
  *= require_self
+ *= require foundation_and_overrides
```
Hmmm ... this seems like a strange ordering.  Were you not able to require foundation, then self, then tree?

rmw commented on app/controllers/application_controller.rb:
```diff
@@ -3,7 +3,9 @@ class ApplicationController < ActionController::Base
+    @current_user ||= User.find_by(id: session[:user_id]) if session[:user_id]
   end
+
+  helper_method :current_user
```
:+1: :hear_no_evil:

rmw commented on app/controllers/maps_controller.rb:
```diff
@@ -0,0 +1,33 @@
+  def index
+  end
+
+  # def search
```
No commented out code in master! :japanese_ogre:

rmw commented on app/controllers/maps_controller.rb:
```diff
@@ -0,0 +1,33 @@
+    geometry = params['features'][0]['geometry']
+    if current_user
+      interests = current_user.interests
+      @attractions = Attraction.search_within(geometry, interests)
```
You could get this down to one line by using `try` and handling `nil` interests in the `search_within` method.

``` ruby
@attractions = Attraction.search_within(geometry, current_user.try(:interests))
```

rmw commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,26 @@
+  end
+
+  def create
+    @user = User.new(user_params)
```
Line break here.

rmw commented on app/models/attraction.rb:
```diff
@@ -0,0 +1,41 @@
+  end
+
+  def self.search_within(geometry, interests = [])
+    Attraction.where("longlat.coordinates" => {
```
you shouldn't need the `Attraction.` prefix here.

rmw commented on app/models/attraction.rb:
```diff
@@ -0,0 +1,41 @@
+  validates_presence_of :name, :longlat, :yelp_id
+  validates_uniqueness_of :yelp_id
+
+  def longitude
```
:+1: to these nice methods!

But you shouldn't need to use `self` for reading.

rmw commented on app/models/attraction.rb:
```diff
@@ -0,0 +1,41 @@
+                      "$geoWithin" => {
+                          "$geometry" => geometry
+                        }
+                      }).all.to_a.select { |attraction| (attraction.categories & interests).any? }
```
Eeek.  `select` is an enumerable method - Is there a way to run that categories and interests check in mongo instead?  (I would think there must be ...)

rmw commented on app/models/location.rb:
```diff
@@ -0,0 +1,21 @@
+
+  validates_presence_of :city, :region, :longlat
+
+  def latitude
```
Since both `Attraction` and `Location` have `latitude` and `longitude` methods, this is the perfect (and really only case) for using a `concern.`

You might have `app/models/concerns/long_lat.rb`

``` ruby
module LongLat
  def latitude
    longlat['coordinates'][1]
  end

  ...
end
```

Then in `Attraction` and `Locations` you'd just `include LongLat`

rmw commented on app/views/attractions/create.js.erb:
```diff
@@ -0,0 +1 @@
+console.log('successful create')
```
Since not all browsers support `console.log`, you should really not use it in production.  

If you must, make you sure define `console.log` for browsers that don't have it: `console = console || {}; console.log = console.log || function() {};`.

rmw commented on app/views/index.html.erb:
```diff
@@ -0,0 +1,11 @@
+<div class="home_container">
+  <div class="get_route">
+  <% form_for :maps, :url => {action: 'directions', controller: 'maps'}, method: 'GET', id:"calculate-route" do |f| %>
+  <%=f.text_area :origin, :required => true, :placeholder => "Origin" %>
```
indentation here!

rmw commented on app/views/layouts/_messages.html.erb:
```diff
@@ -0,0 +1,9 @@
+<%# Rails flash messages styled for Zurb Foundation %>
+<% flash.each do |name, msg| %>
+  <% if msg.is_a?(String) %>
```
When is msg not a string?

rmw commented on app/views/layouts/_messages.html.erb:
```diff
@@ -0,0 +1,9 @@
+<%# Rails flash messages styled for Zurb Foundation %>
+<% flash.each do |name, msg| %>
+  <% if msg.is_a?(String) %>
+    <div data-alert class="alert-box round <%= name.to_s == 'notice' ? 'success' : 'alert' %>">
```
Getting the correct class name is a great job for a helper method!

rmw commented on app/views/maps/start.html.erb:
```diff
@@ -0,0 +1,11 @@
+
+  <div class="small-4 small-centered columns get_route">
+    <%= form_for :maps, :url => {action: 'directions', controller: 'maps'}, method: 'GET', id:"calculate-route" do |f| %>
+    <%=f.text_field :origin, :required => true, :placeholder => "Origin" %>
```
indent under the form!

rmw commented on config/initializers/cookies_serializer.rb:
```diff
@@ -1,3 +1,3 @@
 
-Rails.application.config.action_dispatch.cookies_serializer = :json
\ No newline at end of file
+# Rails.application.config.action_dispatch.cookies_serializer = :json
```
Why did you comment this out?

rmw commented on config/initializers/google.rb:
```diff
@@ -0,0 +1 @@
+GOOGLE = YAML.load_file('config/google.yml')[Rails.env].with_indifferent_access['google_key'] if Rails.env.development?
```
When you are creating global constants, it's preferrable to be very specific about what they store to avoid namespace classes (like a Google gem).  I'd go back to `GOOGLE_KEY`.  Same with Yelp below.

rmw commented on config/routes.rb:
```diff
@@ -1,10 +1,24 @@
+
+  resources :home
+
+  get '/users/:id/categories' => 'users#travel_profile', as: :travel_profile
```
You can do this inside the `resources :users` block [using member/collection](http://guides.rubyonrails.org/routing.html#adding-more-restful-actions).

``` ruby
resource :users do
  get 'categories', to: 'users#travel_profile`, on: :member, as: :travel_profile
end
```

rmw commented on config/routes.rb:
```diff
@@ -1,10 +1,24 @@
+  get 'maps/start' => 'maps#start'
+  get 'maps/directions' => 'maps#directions'
+  post 'maps/search' => 'maps#search'
+  resources :maps, only: [:index]
```
Again, you can move start, directions, and search into the resources: :exclamation: 

``` ruby
resources :maps, only: [:index] do
  collection do
    get 'start'
    get 'directions'
    get 'search'
  end
end
```

rmw commented on db/seeds.rb:
```diff
@@ -5,3 +5,108 @@
+      unless Seed.find_by(location: location, category: category)
+        puts "Searching for #{category.subcategory_code} in #{location.city}, #{location.region}"
+
+        result = Yelp.client.search_by_coordinates(
```
This is _a lot_ of logic to have in your seed file, especially logic that is not tested.  This is a good use case for a service object, such as `AttractionImporter` to handle searching yelp and importing those things into the database.  You also might consider then adding logic that checks if there already is data for that category and skipping it, which would allow you to incrementally add more categories without having to rebuild the database.

rmw commented on spec/models/location_spec.rb:
```diff
@@ -0,0 +1,9 @@
+  it { is_expected.to validate_presence_of(:city) }
+  it { is_expected.to validate_presence_of(:region) }
+  it { is_expected.to validate_presence_of(:longlat) }
+  it { is_expected.to have_many(:attractions) }
```
I would add a test for the `latitude` and `longitude` methods here.  It seems trival but, as you know, it's easy to get confused when you're just dealing with multi-dimensional arrays. :confused:

rmw commented on spec/spec_helper.rb:
```diff
@@ -0,0 +1,89 @@
+require 'simplecov'
+SimpleCov.start
```
You should move `SimpleCov` into `rails_helper` and call `SimpleCov.start 'rails'` to ensure you'd not over inflating your testing.  We wouldn't want that, would we? :wink:

rmw commented:
Holla @gmmowry @nelltaylor @nicole-a-tesla @lduan 
cc @m3rabb @dukegreene 

You've done a kick ass :facepunch: job on some really complex problems! :confetti_ball: You're code is generally really well organized and encapsulated! :tada: 

Other than some lack of indentation in view files and the need for more new lines, the main thing for improvement is ...

### OOJS

You are actually really close with a number of bits already in objects.  

#### App object

It seems to be that the main place you are struggling is setting up the interactions between those objects.  This is an excellent choice for a controller-type object! MVC ftw! :trophy: 

A common pattern in apps today is to have a global `App` object literal:

``` javascript
var App = App || {};

App.init = function() { 
  //create and possibly store objects such as Route and directionsService
  //connect objects together by passing them to an object on creation or via an `init` method
};

$(document).ready(function() { 
  App.init();
});
```

One way to make this easier is to only store variables in a constructor function and then use an `init` method to set up other functionality and/or store other objects.

#### View objects

Separating out the data-holding logic from the data-displaying logic can be very useful in JS (partly because accessing the DOM is one of the most expensive things you can do).  That said, it's often counter-intuitive.

The View object (`RouterView`) holds the selector and DOM element as well as a data/model object (`Router`).  The view will set up click events and pass data to the `Router` object which knows nothing about the DOM, but only about the data it needs.

### Testing

#### Controllers

While you have pretty good model test coverage, I'd like to see tests for the controllers, especially because you're app is basically acting as an API for the front-end making clarity about the interface very important.

#### JavaScript

Since most of the heavy lifting in this app is happening in JS, I'd really really like to see some Jasmine tests.  Refactoring the code into objects will help tremendously in terms of being able to test your JS.  Take a look at [teaspoon](https://github.com/modeset/teaspoon) and [jasmine-jquery matchers](https://github.com/velesin/jasmine-jquery).

Remember that the less you are dependent on `$(document).ready` (which is useless in Jasmine testing) the easier your JS will be to test ... and the more awesome it will be. :boom:

nelltaylor committed:
Merge pull request #31 from lduan/master

Fixed api keys in heroku

nelltaylor committed:
Revert "Fixed api keys in heroku"

gmmowry committed:
Merge pull request #32 from bison-2014/revert-31-master

Revert "Fixed api keys in heroku"

nicole-a-tesla committed:
routes recalculate on the addition of waypoints as selected from the list below the map.

ToDo:
- clear drawn polygon before redraw
- see if ajax can be cooerced into adding waypoints faster
- fix positioning issue
- add more info to li elements

gmmowry committed:
Revert "Revert "Fixed api keys in heroku""

lduan committed:
Merge pull request #33 from bison-2014/revert-32-revert-31-master

Revert "Revert "Fixed api keys in heroku""

lduan committed:
moved api_keys_developement.yml.example file to the correct folder

gmmowry committed:
Merge pull request #34 from lduan/master

moved api_keys_developement.yml.example file to the correct folder

nicole-a-tesla committed:
changes to map.js for waypoints

nelltaylor committed:
Merge pull request #35 from nicole-a-tesla/master

changes to map.js for waypoints

nicole-a-tesla committed:
waypoints dynamically added, polygon clears and redraws with each addition. Woo!

nicole-a-tesla committed:
removed old attractions.js file

gmmowry committed:
debugging

nicole-a-tesla committed:
Merge pull request #10 from nicole-a-tesla/waypoints!

routes recalculate on the addition of waypoints as selected from the list

gmmowry committed:
Merge pull request #36 from nicole-a-tesla/master

waypoints dynamically add on button click, polygon clears and redraws

nelltaylor committed:
Set up variables to pass origin and destination to map display page; renamed custom map routes.

nelltaylor committed:
Added start_trip method to maps controller

gmmowry committed:
Merge pull request #37 from nelltaylor/master

Origin and destination submit from home page finally working

lduan committed:
created a profile page that allows users to change their password as well as listing their interests

lduan committed:
added error messages to the log in and sign up pages

gmmowry committed:
Merge pull request #38 from lduan/master

Created profile page, and added error handling to the sessions/user views.

gmmowry committed:
fixed error with multiple attraction objects being added on one click

nelltaylor committed:
Merge pull request #39 from gmmowry/debugging

Debugged attractions.js file

nelltaylor committed:
Added jQuery for waypoints on start page

gmmowry committed:
Delete .DS_Store

This is a stupid file that does stupid things.

lduan committed:
Update mongo uri

lduan committed:
removed optionalable option in mongoid.yml config

nicole-a-tesla committed:
marker colors change on click, list items jump to top on click

nelltaylor committed:
Added styling css partials and jquery to add waypoints on main page. Waypoints submit does not currently work, waiting on waypoint saving logic.

nicole-a-tesla committed:
if you click on a marker, it becomes purple and list item jumps to top. sam if you lick on list item.

Marker is now a class in seperate file

nicole-a-tesla committed:
Merge pull request #11 from nicole-a-tesla/jquery-side-list

Jquery side list

nelltaylor committed:
Merge pull request #40 from nicole-a-tesla/master

jquery/js for list and markers

nelltaylor committed:
Nav styling in progress

nelltaylor committed:
Making views consistent for map.js testing

nelltaylor committed:
Merged in changes to interact with points on map

nicole-a-tesla committed:
Merge pull request #41 from nelltaylor/master

Styling begun and added jquery for form submit on start page

nicole-a-tesla committed:
route.waypts wasn't setting because in the route's calculateRoute method, this.waypts had been changed to this.pitstop. changed it back

nicole-a-tesla committed:
Merge pull request #12 from nicole-a-tesla/fixing-waypts-not-setting-bug

Fixing waypts not setting bug

gmmowry committed:
Merge pull request #42 from nicole-a-tesla/master

Fixed waypts not setting bug

gmmowry committed:
added some trips functionality

lduan committed:
added a way to persist a list of all attractions already added.  This allows us to not add the same attraction twice.  I also wrapped every attraction in an Attration object, which stores the original attraction data in attrData.  The attraction object also has a reference to its marker (this.marker) and list item DOM element (this.)

gmmowry committed:
Merge pull request #43 from lduan/master

Mongo uri changes + Refactor of javascript to store waypoints and attractions in retrievable variables.

gmmowry committed:
edited maps js file

gmmowry committed:
added trips js file to handle saving user trips

lduan committed:
fix for add to trip

gmmowry committed:
trips controller created to handle trip saving

gmmowry committed:
saving user trips

gmmowry committed:
fixed associations and fields

gmmowry committed:
added save my trip button back

gmmowry committed:
updated interests partial to display subcategories and slideToggle

gmmowry committed:
added partial to view saved trips for user

gmmowry committed:
added partial for trips"
git status

gmmowry committed:
added routes for trips

gmmowry committed:
updated broken yelp csv file

nicole-a-tesla committed:
Merge pull request #44 from lduan/master

fix for add to trip

gmmowry committed:
Removed unnecessary code

gmmowry committed:
Removed method that wasn't being used

nicole-a-tesla committed:
Merge pull request #45 from gmmowry/master

trips can be saved

gmmowry committed:
trips saving route waypoints and yelp waypoints as two separate objects, added yelp_waypoints attribute to trip

gmmowry committed:
making sure push went through

lduan committed:
added remove button to trip list items

nelltaylor committed:
Merge pull request #46 from gmmowry/master

Trips saving google route waypoints and yelp attraction waypoints

nelltaylor committed:
Delete .DS_Store

lduan committed:
added distance and duration to the trip

gmmowry committed:
Merge pull request #47 from lduan/master

Additional functionality (remove items from list and added distance/duration)

nelltaylor committed:
Added precompiling configs to production.rb

nelltaylor committed:
Reconciled styling, again

nelltaylor committed:
Styling changes reconciled mostly with javascript updates; much more styling to come.

lduan committed:
fixed broken waypoint removal

lduan committed:
Merge pull request #48 from nelltaylor/master

All styling reconciled with javascript

nelltaylor committed:
Home page style finalized for mobile, tablet, and desktop. All styling bugs fixed on this page.

nelltaylor committed:
Added constants and set main colors as resuable variables to override Foundation defaults

nicole-a-tesla committed:
Merge pull request #49 from nelltaylor/master

Finalized styles for home page, fixed styling-related bugs for all media on this page

lduan committed:
fixed pitstops, and refactored waypoint code a bit to make it easier to add waypoints.  From now on, you just have to call route.addWaypoint(attraction) or route.removeWaypoint(attraction), and it will handle all of it's internal requests.  No more manually push/popping from this.waypts.

gmmowry committed:
Merge pull request #50 from lduan/master

fixed pitstops

lduan committed:
fixed add waypoint button on the start page.  It now appears again when you add it back in

nicole-a-tesla committed:
fixed the minutes calculation

nicole-a-tesla committed:
Merge pull request #51 from lduan/master

fixed add waypoint button on the start page.  It now appears again when you add it back in.

nelltaylor committed:
Fixed front page not filling full height for large screen displays, began map_display list layout

nicole-a-tesla committed:
Merge pull request #13 from nicole-a-tesla/fix-minutes

Fix minutes

gmmowry committed:
yelp_waypoints now saving as attractions to a trip

gmmowry committed:
added association between attractions and trips

gmmowry committed:
fixed assoc between trips and attractions

nicole-a-tesla committed:
Merge pull request #52 from gmmowry/master

Trips now have saved yelp_waypoints that are Attraction objects

gmmowry committed:
added demo cities and categories for friday

nelltaylor committed:
Merge pull request #53 from nicole-a-tesla/master

fixed minutes display

nelltaylor committed:
Added scrolling to attractions list on map display, working for both desktop and mobile

lduan committed:
added filtering code so that results are filtered by rating, and no more than 5 appear per city

lduan committed:
clicking on a list item now colors and pans its corrisponding marker on the map

lduan committed:
removing a list item is not animated

gmmowry committed:
Merge pull request #54 from lduan/master

Added route filter + clicking on a list item colors the map.

lduan committed:
adding an attraction to the trip now removes the marker from GMaps, but leaves the default waypoint marker.  Originally there were two, which was confusing from a UI standpoint

gmmowry committed:
added city and state fields for attractions

gmmowry committed:
slimmed down category csv for demo

gmmowry committed:
changed seeds file to pull from demo csv, edited dummy user, and added city and statecode to attraction

gmmowry committed:
Merge pull request #55 from lduan/master

minor UI tweaks.  Clicking on an X is now animated and fixed a bug with duplicate markers when an attraction was set as a waypoint.

nelltaylor committed:
Changing Css for saved state of attraction list elements

nelltaylor committed:
Saved state working for adding attractions

gmmowry committed:
Merge pull request #56 from nelltaylor/master

Styling map interaction states

gmmowry committed:
added the city/state location and yelp rating for each attraction

nelltaylor committed:
Removed password management from User/Travel Profile page

gmmowry committed:
Merge pull request #57 from nelltaylor/master

Removed password management from User/Travel Profile page

gmmowry committed:
added yelp category, city/state, and ratings to each attraction listing

lduan committed:
Merge pull request #58 from gmmowry/master

Added more information for attraction listing, dump from earlier is in here too

nelltaylor committed:
JQuery on login and signup working, need to style

nelltaylor committed:
Styling added for forms on home page

nelltaylor committed:
Fixed attractions list gutters

nelltaylor committed:
Added styling to toggle on interests page

gmmowry committed:
Merge pull request #59 from nelltaylor/master

JQuery on login and signup working, need to style

nelltaylor committed:
Added better spacing for subcategories

nelltaylor committed:
Edited city/state display for attractions list

gmmowry committed:
Merge pull request #60 from nelltaylor/master

Edited attraction.js for displaying city and state/category info

nelltaylor committed:
Made subcategories into columns

lduan committed:
renamed app to ReTour in the html title

nelltaylor committed:
Added styling to map and finished basic user profile page

nelltaylor committed:
Fixing small styling bugs, added toggling ui carets on interests

nelltaylor committed:
Merge pull request #61 from lduan/master

renamed app to ReTour in the html title

gmmowry committed:
Merge pull request #62 from nelltaylor/master

Added map styling, added ui carets for smoother toggling on interests, fixed small bugs

nelltaylor committed:
Fixed formatting on trips

gmmowry committed:
Merge pull request #63 from nelltaylor/master

Fixed formatting on trips
